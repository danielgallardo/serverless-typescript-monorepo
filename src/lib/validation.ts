import env from 'env-var';
import _Joi, {Root, SchemaLike, StringSchema, ValidationOptions} from 'joi';
import {JoiError} from './errors';
import {errorToJSON, logDebugInfo, logError, logInfo} from './logger';

export interface IJoi extends Root {
  /**
   * Use this validation for queryParameters and pathParameters
   * to avoid undefined and null strings passed from front-end
   */
  param(): this;
  /**
   * Use this validation for unix timestamps in milliseconds
   * it will not convert value to date
   * use Joi.date().timestamp() to automatically convert value to date
   */
  timestamp(): this;
}

export interface ValidationConfig extends ValidationOptions {
  internal?: boolean;
}

export const Joi = _Joi.extend([
  {
    name: 'param',
    base: _Joi
      .string()
      .not('undefined', 'null')
      .required()
  },
  {
    name: 'timestamp',
    base: _Joi
      .date()
      .timestamp()
      .raw()
  }
]) as IJoi;

/**
 * Attempts to return a valid value, if value is invalid throws a JoiError
 * @param value - a value to validate
 * @param schema - Joi schema
 * @param errorPrefix - a sting that is prepended to error message
 * @param options - Joi validation options
 */
export function attempt<T>(
  value: T,
  schema: SchemaLike,
  errorPrefix?: string,
  options: ValidationConfig = {}
): T {
  const {internal, ...joiOptions} = options;
  const result = Joi.validate<T>(value, schema, joiOptions);
  if (result.error) {
    throw new JoiError(result.error, {errorPrefix, internal});
  }
  return result.value;
}

/**
 * Asserts that value is valid, if value is invalid throws a JoiError
 * @param value - a value to validate
 * @param schema - Joi schema
 * @param errorPrefix - a sting that is prepended to error message
 * @param options - Joi validation options
 */
export function assert<T>(
  value: T,
  schema: SchemaLike,
  errorPrefix?: string,
  options: ValidationConfig = {}
): void {
  const {internal, ...joiOptions} = options;
  const result = Joi.validate<T>(value, schema, joiOptions);
  if (result.error) {
    throw new JoiError(result.error, {errorPrefix, internal});
  }
}

/**
 * Gets and validates process.env variable
 * @param name
 * @param defaultValue
 */
export function getEnv(name: string, defaultValue?: string): string {
  try {
    if (defaultValue) return env.get(name, defaultValue).asString();
    return env
      .get(name)
      .required()
      .asString();
  } catch (error) {
    // we need to log an error right away in case getEnv is used outside lambda handler
    // otherwise we'll get only "module initialization error: EnvVarError"
    console.log(`== MISSING ENV VARIABLE: "${name}" ==`, JSON.stringify(errorToJSON(error)));
    throw error;
  }
}
