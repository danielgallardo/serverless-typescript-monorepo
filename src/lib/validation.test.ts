import {assert, attempt, getEnv, Joi} from './validation';

// define env variables for your test at the top
process.env.AWS_REGION = 'eu-west-1';

const schema = Joi.object().keys({
  foo: Joi.string().required()
});

describe('validation', () => {
  it('attempt should return correct value', () => {
    expect(attempt({foo: 'bar'}, schema)).toEqual({foo: 'bar'});
  });

  it('attempt throw error if value is incorrect', () => {
    expect(attempt.bind(null, {foo: 1}, schema, 'Validation test')).toThrowErrorMatchingSnapshot();
  });

  it('assert should not throw error if value is valid', () => {
    expect(assert.bind(null, {foo: 'bar'}, schema)).not.toThrow();
  });

  it('assert throw error if value is incorrect', () => {
    expect(assert.bind(null, {}, schema, 'Validation test')).toThrowErrorMatchingSnapshot();
  });

  it('getEnv should return env variable', () => {
    expect(getEnv('AWS_REGION')).toBe('eu-west-1');
    expect(getEnv('MISSING_VAR', 'foo')).toBe('foo');
  });

  it('getEnv should throw error if variable is not defined', () => {
    const spy = jest.spyOn(console, 'log');
    expect(getEnv.bind(null, 'MISSING_VAR')).toThrowErrorMatchingSnapshot();
    expect(spy.mock.calls[0][0]).toBe('== MISSING ENV VARIABLE: "MISSING_VAR" ==');
  });

  it('Joi.param should correctly validate query and path parameters', () => {
    const schema = Joi.object().keys({
      userId: Joi.param()
    });
    expect(assert.bind(null, {userId: 'undefined'}, schema)).toThrowErrorMatchingSnapshot();
    expect(assert.bind(null, {userId: 'null'}, schema)).toThrowErrorMatchingSnapshot();
  });
});
