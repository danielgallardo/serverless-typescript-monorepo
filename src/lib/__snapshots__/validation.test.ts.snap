// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation Joi.param should correctly validate query and path parameters 1`] = `"\\"userId\\" contains an invalid value"`;

exports[`validation Joi.param should correctly validate query and path parameters 2`] = `"\\"userId\\" contains an invalid value"`;

exports[`validation assert throw error if value is incorrect 1`] = `"Validation test: \\"foo\\" is required"`;

exports[`validation attempt throw error if value is incorrect 1`] = `"Validation test: \\"foo\\" must be a string"`;

exports[`validation getEnv should throw error if variable is not defined 1`] = `"env-var: \\"MISSING_VAR\\" is a required variable, but it was not set"`;
